src/main/java/teammates/common/util/Url.java

```diff
@@ -1,15 +1,17 @@
 package teammates.common.util;
 
-public class Url {
+/**
+ * The Url class represents a URL string.
+ * It provides methods to manipulate the URL string and extract values from it.
+ */
+public abstract class Url {
```

Comment:
Here's the API I propose for the Url class:

* **`Url(String fullUrl)`**
* **`Url(String baseUrl, String relativePath)`** 
  * baseUrl is of the form `scheme:[//[user:password@]host[:port]][/]`. e.g. `http://google.com` `https://localhost:8888/`
  * relativePath is of the form `[/]path[?query][#fragment]`. Trailing slashes will be ignored.
* **`setRelativePath(String relativePath)`** : sets the relative path of the object.
  * relativePath is of the form ...
* **`getRelativePath():String`**: no leading or trailing slashes
* **`setBaseUrl(String: baseUrl)`**
* **`getBaseUrl():String`**: no trailing slash
* **`toString():String`** : returns the full Url, no trailing slash

When somebody asks why we designed it that way, the answer is "because that's how URL's work", rather than "that's the API that fits our use cases best". The former requires the reader to know how URLs work only, while the latter requires the reader to know how URLs work, our use cases, and how our API is different from what is normally expected of a URL class.

Now that we have come this far in this discussion, may be we should also explore another alternative involving the *Adapter pattern*. Instead of AppUrl extending our own Url class, we can make AppUrl wrap around `java.net.URL` (i.e. internally it keeps a `java.net.URL` object.